toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1388 @@
+{
+  // The path on the file system of the watched resource.
+  nsString mPath;
+
+  // The watched descriptor returned from inotify_add_watch().
+  int mWatchedResourceDescriptor;
+
+  WatchedResourceDescriptor(const nsAString& aPath, const int anHandle)
+    : mPath(aPath)
+    , mWatchedResourceDescriptor(anHandle)
+  {
+  }
+};
+
+/**
+ *  * A structure used to pass the callbacks to the AddPathRunnableMethod() and
+ *   * RemovePathRunnableMethod().
+ *    */
+struct PathRunnablesParametersWrapper
+{
+  nsString mPath;
+  nsMainThreadPtrHandle<nsINativeFileWatcherCallback> mChangeCallbackHandle;
+  nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback> mErrorCallbackHandle;
+  nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>
+    mSuccessCallbackHandle;
+
+  PathRunnablesParametersWrapper(
+    const nsAString& aPath,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>&
+      aOnSuccess)
+    : mPath(aPath)
+    , mChangeCallbackHandle(aOnChange)
+    , mErrorCallbackHandle(aOnError)
+    , mSuccessCallbackHandle(aOnSuccess)
+  {
+  }
+};
+
+/**
+ * This runnable is dispatched from the main thread to get the notifications of
+ * the changes in the watched resources by continuously calling the blocking
+ * function GetQueuedCompletionStatus. This function queries the status of the
+ * Completion I/O port initialized in the main thread. The watched resources are
+ * registered to the completion I/O port when calling |addPath|.
+ *
+ * Instead of using a loop within the Run() method, the Runnable reschedules
+ * itself by issuing a NS_DispatchToCurrentThread(this) before exiting. This is
+ * done to allow the execution of other runnables enqueued within the thread
+ * task queue.
+ */
+class NativeFileWatcherIOTask : public Runnable
+{
+public:
+  explicit NativeFileWatcherIOTask()
+    : Runnable("NativeFileWatcherIOTask")
+    , mShuttingDown(false)
+  {
+  }
+
+  NS_IMETHOD Run() override;
+  nsresult AddPathRunnableMethod(
+    PathRunnablesParametersWrapper* aWrappedParameters);
+  nsresult RemovePathRunnableMethod(
+    PathRunnablesParametersWrapper* aWrappedParameters);
+  nsresult DeactivateRunnableMethod();
+
+private:
+  // Maintain 2 indexes - one by resource path, one by resource |HANDLE|.
+  // Since |HANDLE| is basically a typedef to void*, we use nsVoidPtrHashKey to
+  // compute the hashing key. We need 2 indexes in order to quickly look up the
+  // changed resource in the Worker Thread.
+  // The objects are not ref counted and get destroyed by
+  // mWatchedResourcesByPath on NativeFileWatcherService::Destroy or in
+  // NativeFileWatcherService::RemovePath.
+  nsClassHashtable<nsStringHashKey, WatchedResourceDescriptor>
+    mWatchedResourcesByPath;
+  nsDataHashtable<nsVoidPtrHashKey, WatchedResourceDescriptor*>
```

Comment:
Oh, you haven't changed it. I'm nearly sure that `nsVoidPtrHashKey` is the wrong choice under Linux. You probably want `nsGenericHashKey<int>`.